### PR TITLE
Add caching for product attributes used in sales rules

### LINKED ISSUE #33900: but to fix https://github.com/magento/magento2/pull/33537
### Preconditions (*)

1.
2.

### Steps to reproduce (*)

1.
2.

### Expected result (*)

1. [Screenshots, logs or description]
2.

### Actual result (*)

1. [Screenshots, logs or description]
2.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by ihor-sviziev**
### Description (*)
This PR improves response time & reduces load to DB while fetching list of carts using WebAPI:
```bash
curl --location -g --request GET 'https://www.magento24.test/index.php/rest/default/V1/carts/search?searchCriteria[filter_groups][0][filters][0][field]=updated_at&searchCriteria[filter_groups][0][filters][0][condition_type]=gt&searchCriteria[sortOrders][0][field]=updated_at&searchCriteria[sortOrders][0][direction]=DESC&searchCriteria[currentPage]=1&searchCriteria[pageSize]=50&searchCriteria[filter_groups][0][filters][0][value]=2021-06-25+12%3A00%3A00' \
--header 'Authorization: Bearer ibvfeew3inotn2u1gu573qd3ndbunu8k'
```

Here is our performance difference, while we have quite a lot of cart price rules (and the result is just 10 different attributes):
image

You might have different results depending on the amount of cart price rules and user attributes in it.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Unit Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

File: app/code/Magento/SalesRule/Model/Plugin/QuoteConfigProductAttributes.php [hunk 1]
```diff
@@ -15,31 +17,50 @@ class QuoteConfigProductAttributes
...
+     * @return array
+     */
+    private function getActiveAttributes(): array
+    {
+        if ($this->activeAttributes === null) {
+            $this->activeAttributes = $this->ruleResource->getActiveAttributes();
+        }
+
+        return $this->activeAttributes;
+    }
```
Comment (Den4ik) [inline app/code/Magento/SalesRule/Model/Plugin/QuoteConfigProductAttributes.php]:
```suggestion
    private function getActiveAttributeCodes(): array
    {
        if ($this->activeAttributeCodes === null) {
            $this->activeAttributeCodes = array_column($this->ruleResource->getActiveAttributes(), 'attribute_code');
        }

        return $this->activeAttributeCodes;
    }
```

File: app/code/Magento/SalesRule/Model/Plugin/QuoteConfigProductAttributes.php [hunk 2]
```diff
@@ -15,31 +17,50 @@ class QuoteConfigProductAttributes
     /**
      * @var RuleResource
      */
-    protected $_ruleResource;
+    private $ruleResource;
+
+    /**
+     * @var array|null
+     */
+    private $activeAttributes;
```
Comment (Den4ik) [inline app/code/Magento/SalesRule/Model/Plugin/QuoteConfigProductAttributes.php]:
```suggestion
    private $activeAttributeCodes;
```

File: app/code/Magento/SalesRule/Model/Plugin/QuoteConfigProductAttributes.php [hunk 3]
```diff
@@ -15,31 +17,50 @@ class QuoteConfigProductAttributes
...
+    public function afterGetProductAttributes(Config $subject, array $attributeKeys): array
     {
-        $attributes = $this->_ruleResource->getActiveAttributes();
+        $attributes = $this->getActiveAttributes();
+
         foreach ($attributes as $attribute) {
             $attributeKeys[] = $attribute['attribute_code'];
         }
+
         return $attributeKeys;
```
Comment (Den4ik) [inline app/code/Magento/SalesRule/Model/Plugin/QuoteConfigProductAttributes.php]:
```suggestion
        return array_merge($attributeKeys, $this->getActiveAttributeCodes());
```

**Review by Den4ik — CHANGES_REQUESTED** [review]
Hi @ihor-sviziev. I believe that we can little bit improve changes with caching attribute codes

File: app/code/Magento/SalesRule/Model/Plugin/QuoteConfigProductAttributes.php [hunk 4]
```diff
@@ -15,31 +17,50 @@ class QuoteConfigProductAttributes
...
      * @param array $attributeKeys
      *
      * @return array
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
-    public function afterGetProductAttributes(\Magento\Quote\Model\Quote\Config $subject, array $attributeKeys)
+    public function afterGetProductAttributes(Config $subject, array $attributeKeys): array
     {
-        $attributes = $this->_ruleResource->getActiveAttributes();
+        $attributes = $this->getActiveAttributes();
```
Comment (ihor-sviziev) [inline app/code/Magento/SalesRule/Model/Plugin/QuoteConfigProductAttributes.php:46]:
```suggestion
        $attributes = $this->getActiveAttributeCodes();
```

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests EE, Unit Tests

**Review by Den4ik — APPROVED** [review]
✅ Approved.
@ihor-sviziev Thanks for the work

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9159](https://jira.corp.magento.com/browse/ENGCOM-9159) has been created to process this Pull Request
:eight_spoked_asterisk: @Den4ik, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
@magento create issue

**Comment by engcom-Alfa** [comment]
@magento create issue

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have a Magento 2 instance with sample data.
2. It would be good, if we have got a few days old instance in which we have already performed some activity.



**Manual testing scenario:**

1. Get the admin token using POST request /rest/default/V1/integration/admin/token to collect authorization token

2. Request a GET call, utilizing a above Authorization code using URI http://local.magento2.com/index.php/rest/default/V1/carts/search?searchCriteria[filter_groups][0][filters][0][field]=updated_at&searchCriteria[filter_groups][0][filters][0][condition_type]=gt&searchCriteria[sortOrders][0][field]=updated_at&searchCriteria[sortOrders][0][direction]=DESC&searchCriteria[currentPage]=1&searchCriteria[pageSize]=50&searchCriteria[filter_groups][0][filters][0][value]=2021-06-25
 
3. Validate the GET response time



**Before:** :heavy_multiplication_x: The response time was more comparatively.

image


**After:** :heavy_check_mark:  The response time has been reduced 25%

image


Even there is no huge difference in the response load time before and after fix, the given fix at last reduces the response time certain extent and has no impact on the response data.
There is no additional regression required in this case since it is related to the WebAPI request load time.

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
Hi @sidolov When it could be merged?

**Pull request merged**